### ISSUE TITLE
[lab1] 回归评测函数实现&分类模型实现 的几个问题

### ISSUE THREAD

**Issue by w4ynex** [Smadx/USTC-ML24-Fall#20]
1. 想确认一下 任务1.6 中要求完成的  eval_LinearRegression() 函数是要从 TestLoader 中加载出所有的 batch 还是一个 batch 计算误差？我认为实验文档表达的是用 for batch in loader 计算出所有测试集中数据的预测均值；
2. 关于调参：在 任务1.2 中，我们把训练集和验证集都加载到 TrainLoader 用于训练，我们应该根据 eval 的结果进行参数调整吗？这样的话实验里的 TestLoader 对应的数据就相当于验证集吗？
3. 分类任务2.1(b) 划分数据集：我认为 2.1(b) 中实现的函数 data_split_classification() 和 1.1(b) 实现的data_split_regression() 功能相同，为什么要实现两个相同功能的函数呢？是我的理解有问题吗？
4. 分类模型的实现： 代码框架要求将分类模型的参数统一注册为 beta（换言之，我理解 beta 是原来权重 w 和偏置 b 合并的结果），那么 beta 的形状是 (in_feature, 1) 还是 (in_feature + 1, 1) 呢？模型接受的输入 x 的 in_feature 包含常数 1 吗（为了达到偏置的效果），还是需要在模型 predict 时人为地为输入 x 加上一列呢？

感谢助教解答！

**Comment by Smadx** [comment]
## 1
计算测试集中的所有数据的预测均值
## 2
不应该,在整个训练过程中测试集只能被使用一次,你需要根据验证集上的结果调整超参数
## 3
分类问题的数据集被重新处理过了,所以要重新划分(白送你分别不要啊)
## 4
beta的形状需要你自己决定；你需要在模型接受输入前给输入人为地加上一列1

**Comment by w4ynex** [comment]
第二点调参还是有点不明白（x）
根据实验文档的要求，data_split_regression() 返回了 train_loader 和 test_loader。其中 train_loader 是包含训练集和验证集的。
感觉把验证集单独拿出来调参需要修改 submisson.py 中已经完成的部分代码。还是说直接用调整超参数，根据训练最后的 loss 选取一组较优的超参数用于评估？

**Comment by Smadx** [comment]
``train_loader``包含的是"当次操作的训练集",在三种情况下分别为:
- 训练集
- 训练集+验证集
- 训练集+验证集+测试集

**Comment by w4ynex** [comment]
一个关于问题 4 的延申：
在完成了 `submission.py` 后我发现 Classification 部分的对代码框架的注释存在不自洽的情况，对整体的实现没有太大影响。
- 在 trainC.py 中我们可以看到 `--in_features` 默认为 14，即没有包含 bias_term。
- 我们在 `model = LogisticRegression(cfg.in_features)` 中将 model 的 in_feature 规定为 14。
- 在 2.6.a `eval_LogisticRegression()` 的注释中写到
```
    # 2.6-a
    # Iterate over the data and compute the accuracy
    # This time, we use the whole dataset instead of a DataLoader.Don't forget to add a bias term to the input
```
注释要求我们将 bias_term 加入 input ，此时 `model.predict()` 的输入的形状为 `(batch_size, in_feature + 1)`。
- `model.predict()` 注释中的示例用法中 model 的输入形状为 `(batch_size, in_feature)` 与前面输入的形状为  `(batch_size, in_feature + 1)` 矛盾。
```
Example::
    
            >>> from model import LogisticRegression
            >>> # Define the model
            >>> model = LogisticRegression(3)
            >>> # Predict
            >>> x = np.random.randn(10, 3)
            >>> y = model(x)
```
___

我认为加入 bias_term 有两种方式：
1. 在 TrainerC 中读取 dataset 时人为向 feature 中加入 bias term；
2. 在 predict 和 loss 的 `backward()` 中为输入的 `x` 手动加入 bias term，虽然感觉性能上略差于第一种方法。

我根据注释中的示例采取了第二种实现方法，保证了 model 的 `in_features` 与输入 `x` 形状的最后一维保持一致。虽然感觉对结果没啥影响，但感觉在实现过程中还是有点 confusing 的。

**Comment by Smadx** [comment]
:+1:
